cloudinit/stages.py

```diff
@@ -702,6 +716,47 @@ def _find_networking_config(self):
         return (self.distro.generate_fallback_config(),
                 NetworkConfigSource.fallback)
 
+    def update_event_enabled(
+        self, event_source_type: EventType, scope: EventScope = None
+    ) -> bool:
+        """Determine if a particular EventType is enabled.
+
+        For the `event_source_type` passed in, check whether this EventType
+        is enabled in the `updates` section of the userdata. If `updates`
+        is not enabled in userdata, check if defined as one of the
+        `default_events` on the datasource. `scope` may be used to
+        narrow the check to a particular `EventScope`.
+
+        Note that on first boot, userdata may NOT be available yet. In this
+        case, we only have the data source's `default_update_events`,
+        so an event that should be enabled in userdata may be denied.
+        """
+        default_events = self.datasource.default_update_events  # type: Dict[EventScope, Set[EventType]]    # noqa: E501
+        user_events = userdata_to_events(self.cfg.get('updates', {}))  # type: Dict[EventScope, Set[EventType]]  # noqa: E501
+        # A value in the first will override a value in the second
+        allowed = util.mergemanydict([
+            copy.deepcopy(user_events),
+            copy.deepcopy(default_events),
+        ])
+        LOG.debug('Allowed events: %s', allowed)
+
+        if not scope:
+            scopes = allowed.keys()
+        else:
+            scopes = [scope]
+        scope_values = [s.value for s in scopes]
+        LOG.debug('Possible scopes for this event: %s', scope_values)
+
+        for evt_scope in scopes:
+            if event_source_type in allowed.get(evt_scope, []):
+                LOG.debug('Event Allowed: scope=%s EventType=%s',
+                          evt_scope.value, event_source_type)
+                return True
+
+        LOG.debug('Event Denied: scopes=%s EventType=%s',
```

Comment:
I'm still not happy with this. We attempt to apply networking config twice on every boot. Pre-this-PR behavior, if only BOOT_NEW_INSTANCE was enabled, we would apply exactly once on the first part of first boot, then deny on every subsequent attempt. If BOOT was enabled, we'd apply every single time (twice per boot, including first boot).

With this PR, on first boot we don't have access to userdata, so we fall back to the default update events of the datasource. If only BOOT_NEW_INSTANCE is enabled, we'll get an event denied message in the logs, even if BOOT is specified in the userdata. I'm concerned having an event denied message in the logs will be confusing given that it is enabled in the userdata.

I could add a check for first boot here and slightly alter the log message. It could be something like "Event Denied from data source...", but that also seems like overkill checking every boot for a not-much-more-helpful log message.

Any thoughts for improvement?

Reply:
Are there cases when we would only successfully apply networking configuration on the second attempt during first boot? If so, then does it follow that `BOOT_NEW_INSTANCE` should be the toggle for both of those runs? And, therefore, should we actually be using the check for first boot for more than just a log message?

Reply:
It also occurs to me that we already have language for these occasions, from our script execution modules: once, instance, and boot. I'm not suggesting we change the language we're using here ("instance" is obviously ~contained within "BOOT_NEW_INSTANCE"), but it does make me wonder if we should be following those modules and tracking whether we have applied network configuration _for this instance_, rather than trying to infer what we should be doing based on the type of a boot we think we have.

If we haven't yet configured networking, `BOOT_NEW_INSTANCE` or `BOOT` will cause it to be applied; if we have, then only `BOOT` will.

What do you think?

Reply:
> Are there cases when we would only successfully apply networking configuration on the second attempt during first boot?

Yes, I've been assuming that in my explanations. We will be in BOOT_NEW_INSTANCE state until the datasource has been initialized. It's entirely possible that the first attempt can result in "wait till networking is up", so the second attempt is still BOOT_NEW_INSTANCE. We always attempt to apply networking when BOOT_NEW_INSTANCE

> If so, then does it follow that BOOT_NEW_INSTANCE should be the toggle for both of those runs? And, therefore, should we actually be using the check for first boot for more than just a log message?

I'm not sure I follow what `should be the toggle` means. But to the second question, I think the answer is no. First boot is really only relevant because we don't have userdata available yet.

> but it does make me wonder if we should be following those modules and tracking whether we have applied network configuration for this instance, rather than trying to infer what we should be doing based on the type of a boot we think we have.
If we haven't yet configured networking, BOOT_NEW_INSTANCE or BOOT will cause it to be applied; if we have, then only BOOT will.

This is essentially what I'm already doing. Or am I misunderstanding what you're saying?

Sorry if I didn't make it clear, the but the case I'm concerned about is *when* networking is applied on first attempt and userdata specifies BOOT (which we can't access on first boot), what's happening on the second attempt.

CURRENT BEHAVIOR:
```
master when datasource specifies BOOT_NEW_INSTANCE:
first boot, one: apply
first boot, two: skip
subsequent boot, one: skip
subsequent boot, two: skip

master when datasource specifies BOOT:
first boot, one: apply
first boot, two: apply
subsequent boot, one: apply
subsequent boot, two: apply
```

The above should be exactly the same on this branch when userdata specifies BOOT_NEW_INSTANCE. However, when userdata specifies BOOT:

BRANCH BEHAVIOR:
```
branch with userdata BOOT:
first boot, one: apply
first boot, two: apply or skip based on datasource default because we don't have userdata
subsequent boot, one: apply
subsequent boot, two: apply
```
(These scenarios are assuming networking gets applied on the first attempt)

If on `first boot, two` we skip because of the datasource default, we get a big ugly "Event Denied" log message that will likely be confusing for somebody that doesn't know userdata won't take effect till next boot.

Changing this to instead apply on both attempts isn't difficult, but it would also apply in the BOOT_NEW_INSTANCE case, which is a change in current behavior. This could be important for backwards compatibility, in case we specifically DON'T want network config applied more than once when `BOOT_NEW_INSTANCE` is specified.

Does my reasoning here make sense? If so, do you think it makes sense to modify the log message on first boot for the case I mentioned?